pkg/apis/batch/validation/validation.go

```diff
@@ -565,6 +567,35 @@ func ValidateJobTemplateSpec(spec *batch.JobTemplateSpec, fldPath *field.Path, o
 	return allErrs
 }
 
+func validateCompletions(spec, oldSpec batch.JobSpec, fldPath *field.Path) field.ErrorList {
+	if !feature.DefaultFeatureGate.Enabled(features.ElasticIndexedJob) {
+		return apivalidation.ValidateImmutableField(spec.Completions, oldSpec.Completions, fldPath)
+	}
+	// Completions is immutable for non-indexed jobs.
+	// For Indexed Jobs, if ElasticIndexedJob feature gate is not enabled,
+	// fall back to validating that spec.Completions is always immutable.
+	var allErrs field.ErrorList
+	isIndexedJob := spec.CompletionMode != nil && *spec.CompletionMode == batch.IndexedCompletion
+	if !isIndexedJob {
+		return apivalidation.ValidateImmutableField(spec.Completions, oldSpec.Completions, fldPath)
+	}
+
+	// Indexed Jobs cannot set completions to nil. The nil check
+	// is already performed in validateJobSpec, no need to add another error.
+	if spec.Completions == nil || oldSpec.Completions == nil {
+		return allErrs
+	}
+
+	// For Indexed Jobs, spec.Completions can only be mutated in tandem
+	// with spec.Parallelism such that spec.Completions == spec.Parallelism
+	if *spec.Completions != *oldSpec.Completions {
+		if *spec.Completions != *spec.Parallelism || *oldSpec.Completions != *oldSpec.Parallelism {
```

Comment:
why fail validation of the new incoming object if `*oldSpec.Completions != *oldSpec.Parallelism`? that seems like it makes it impossible to update to fix data?

Reply:
We had a discussion about this here https://github.com/kubernetes/enhancements/pull/3724#discussion_r1086673278, the argument was having a consistent experience that `completions` is allowed to mutate only when changed in tandem with `parallelism`, hence the before and after update check. I am ok with removing the before check and only validate on the after update.

/cc @alculquicondor

Reply:
Fine with me to only check after update.